### ISSUE TITLE
tsquery.map()'s contract should match the TypeScript visitor's contract

### ISSUE THREAD

**Issue by alonsohki** [phenomnomnominal/tsquery#75]
In TypeScript's visitor, it is possible to return either of the following three values:
- The original node: this will leave the AST intact
- A replacement node: this will replace the currently visited node with that replacement
- `undefined`: this will remove the original node from the AST with no replacement

With the current implementation, when a node transformer returns `undefined`, it will leave the tree as is, which breaks TypeScript's visitor's contract and also brings no additional value to simply returning the original node.

As this would be a **breaking change** with a potential big impact, what I suggest is:
- Introducing a new `visit()` functionality, that will conform to TypeScript's visitor contract.

If the library author is okay with this, I'm also happy to open a PR and contribute this improvement myself.

**Comment by phenomnomnominal** [comment]
Yeah I think this makes sense and honestly I'm fine with the breaking change. More than happy to take a PR

**Comment by alonsohki** [comment]
Hello @phenomnomnominal 

I have submitted a PR (https://github.com/phenomnomnominal/tsquery/pull/76) for this. But I have detected one additional issue in the `tsquery.map()` that I would like to fix in this PR, unless it is actually a feature:

`tsquery.map()` doesn't call the transformer when the targeted node is descendant of a node that also matches the same selector. Take this example: A labeled statement whose statement is another labeled statement. If your selector is `LabeledStatement`, it will trigger for the first node but not for the child one. To reproduce:

```ts
      const sourceFile = ts.createSourceFile('index.ts', 'label1: label2: 1 + 1;', ts.ScriptTarget.Latest, true);
      
      const results1 = tsquery.query(sourceFile, 'LabeledStatement');
      console.log(results1.length); // Will print 2

      let count = 0;
      tsquery.map(sourceFile, 'LabeledStatement', (node) => { ++count; return node; });
      console.log(count); // Will print 1
```

It is pretty easy to fix it, so I will add a commit to my PR to fix the issue. But if this was actually intentional, I will simply drop the commit and force-push to remove it. Cheers!
